### PR TITLE
Add checkpoint deletion feature to SDXL

### PR DESCRIPTION
**Pull request by HaoyangLee**
The sdxl ckpt files are too large (~13GB each) to save in local server especially `args.save_ckpt_interval` is set to be a small value. We can keep the latest `args.max_num_ckpt` ckpt files and delete those old ones. Keep all ckpts when `args.max_num_ckpt` is 0.

### CODE REVIEW

File: examples/stable_diffusion_xl/gm/helpers.py [hunk 1]
```diff
@@ -644,16 +649,45 @@ def save_checkpoint(model, path, only_save_lora=False):
...
+            del_ckpts = [del_ckpt_lora]
+        else:
+            del_ckpts = [del_ckpt, del_ckpt_lora]
+
+        for to_del in del_ckpts:
+            if os.path.isfile(to_del):
+                try:
+                    os.chmod(to_del, stat.S_IWRITE)
+                    os.remove(to_del)
+                    print(
```
Comment (hadipash) [inline examples/stable_diffusion_xl/gm/helpers.py:681]: Use a logger to print out messages. And this message could be a debug message as it's not important.

File: examples/stable_diffusion_xl/gm/helpers.py [hunk 2]
```diff
@@ -644,16 +649,45 @@ def save_checkpoint(model, path, only_save_lora=False):
...
+        del_ckpt_lora = _build_lora_ckpt_path(del_ckpt)
+        if only_save_lora:
+            del_ckpts = [del_ckpt_lora]
+        else:
+            del_ckpts = [del_ckpt, del_ckpt_lora]
+
+        for to_del in del_ckpts:
+            if os.path.isfile(to_del):
+                try:
+                    os.chmod(to_del, stat.S_IWRITE)
```
Comment (hadipash) [inline examples/stable_diffusion_xl/gm/helpers.py:679]: Is it necessary to change the mode of the checkpoint file? It was saved by the same script, so it should be writable by default.

**Review by hadipash — APPROVED** [review]

File: examples/stable_diffusion_xl/gm/helpers.py (hunk 1 above)
Comment (HaoyangLee) [reply examples/stable_diffusion_xl/gm/helpers.py]: Updated

File: examples/stable_diffusion_xl/gm/helpers.py (hunk 2 above)
Comment (HaoyangLee) [reply examples/stable_diffusion_xl/gm/helpers.py]: Updated

File: examples/stable_diffusion_xl/train.py [hunk 3]
```diff
@@ -52,6 +52,12 @@ def get_parser_train():
     parser.add_argument("--save_path_with_time", type=ast.literal_eval, default=True)
     parser.add_argument("--log_interval", type=int, default=1, help="log interval")
     parser.add_argument("--save_ckpt_interval", type=int, default=1000, help="save ckpt interval")
+    parser.add_argument(
+        "--max_num_ckpt",
```
Comment (wtomin) [inline examples/stable_diffusion_xl/train.py:56]: It makes more sense to set the default value as `None`. If not specified, `None` means to keep all ckpts. If specified, it needs to be a positive integer, like this:
```
if args.max_num_ckpt is not None:
    assert args.max_num_ckpt > 0, "max_num_ckpt must be a positive integer!"
```

File: examples/stable_diffusion_xl/gm/helpers.py [hunk 4]
```diff
@@ -644,16 +646,46 @@ def save_checkpoint(model, path, only_save_lora=False):
...
+        path_lora = _build_lora_ckpt_path(path)
         ms.save_checkpoint(ckpt_lora, path_lora)
         print(f"save lora checkpoint to {path_lora}")
 
 
+def delete_checkpoint(ckpt_queue, max_num_ckpt, only_save_lora):
+    """
+    Only keep the latest `max_num_ckpt` ckpts while training. If max_num_ckpt == 0, keep all ckpts.
+    """
+    if max_num_ckpt > 0 and len(ckpt_queue) >= max_num_ckpt:
```
Comment (wtomin) [inline examples/stable_diffusion_xl/gm/helpers.py:665]: Then here can be changed to

**Review by wtomin — APPROVED** [review]

File: examples/stable_diffusion_xl/gm/helpers.py (hunk 4 above)
Comment (HaoyangLee) [reply examples/stable_diffusion_xl/gm/helpers.py]: Updated

File: examples/stable_diffusion_xl/train.py (hunk 3 above)
Comment (HaoyangLee) [reply examples/stable_diffusion_xl/train.py:56]: Updated

File: examples/stable_diffusion_xl/gm/helpers.py [hunk 5]
```diff
@@ -632,7 +630,11 @@ def perform_save_locally(save_path, samples):
         base_count += 1
 
 
-def save_checkpoint(model, path, only_save_lora=False):
+def _build_lora_ckpt_path(ckpt_path):
+    return ckpt_path[: -len(".ckpt")] + "_lora.ckpt"
```
Comment (zhtmike) [inline examples/stable_diffusion_xl/gm/helpers.py:634]: hard to read, just to replace by `ckpt_path.replace(".ckpt", "_lora.ckpt"`)

File: examples/stable_diffusion_xl/gm/helpers.py [hunk 6]
```diff
@@ -644,16 +646,46 @@ def save_checkpoint(model, path, only_save_lora=False):
...
+
+        for to_del in del_ckpts:
+            if os.path.isfile(to_del):
+                try:
+                    os.remove(to_del)
+                    logging.debug(
+                        f"The ckpt file {to_del} is deleted, because the number of ckpt files exceeds the limit {max_num_ckpt}."
+                    )
+                except OSError as e:
+                    logging.error(e)
```
Comment (zhtmike) [inline examples/stable_diffusion_xl/gm/helpers.py:681]: in this case, use `logging.exception` will directly show the tracback

File: examples/stable_diffusion_xl/train.py [hunk 7]
```diff
@@ -194,6 +200,7 @@ def train(args):
         raise ValueError("args.ms_mode value must in [0, 1]")
 
     # 5. Start Training
+    assert args.max_num_ckpt is None or args.max_num_ckpt > 0, "args.max_num_ckpt must be None or a positive integer!"
```
Comment (zhtmike) [inline examples/stable_diffusion_xl/train.py:203]: for best practise, using raise Exception is better than asset

Comment (HaoyangLee) [reply examples/stable_diffusion_xl/train.py]: Updated

File: examples/stable_diffusion_xl/gm/helpers.py (hunk 6 above)
Comment (HaoyangLee) [reply examples/stable_diffusion_xl/gm/helpers.py]: Updated

File: examples/stable_diffusion_xl/gm/helpers.py (hunk 5 above)
Comment (HaoyangLee) [reply examples/stable_diffusion_xl/gm/helpers.py]: Updated

**Review by zhtmike — APPROVED** [review]

**Pull request merged**